File: py_ionization/electron_impact/cross_section.py
from __future__ import annotations

from functools import lru_cache
from importlib.resources import files

import numpy as np

from py_ionization._utils import read_x1e16_table


_SPECIES_TO_FILE = {
    "h": "electron_impact_H.txt",
    "h+": "electron_impact_H.txt",
    "c": "electron_impact_C.txt",
    "c+": "electron_impact_C.txt",
    "o": "electron_impact_O.txt",
    "o+": "electron_impact_O.txt",
    "o2": "electron_impact_O2.txt",
    "o2+": "electron_impact_O2.txt",
    "co": "electron_impact_CO.txt",
    "co+": "electron_impact_CO.txt",
    "n2": "electron_impact_N2.txt",
    "n2+": "electron_impact_N2.txt",
    "co2": "electron_impact_CO2.txt",
    "co2+": "electron_impact_CO2.txt",
    "ar": "electron_impact_Ar.txt",
    "ar+": "electron_impact_Ar.txt",
    "na": "electron_impact_Na.txt",
    "na+": "electron_impact_Na.txt",
    "mg": "electron_impact_Mg.txt",
    "mg+": "electron_impact_Mg.txt",
    "al": "electron_impact_Al.txt",
    "al+": "electron_impact_Al.txt",
    "si": "electron_impact_Si.txt",
    "si+": "electron_impact_Si.txt",
    "k": "electron_impact_K.txt",
    "k+": "electron_impact_K.txt",
    "ca": "electron_impact_Ca.txt",
    "ca+": "electron_impact_Ca.txt",
    "ti": "electron_impact_Ti.txt",
    "ti+": "electron_impact_Ti.txt",
    "fe": "electron_impact_Fe.txt",
    "fe+": "electron_impact_Fe.txt",
}


def _data_file(species: str) -> str:
    key = str(species).lower().strip().replace(" ", "")
    try:
        return _SPECIES_TO_FILE[key]
    except KeyError as exc:
        supported = ", ".join(sorted(set(_SPECIES_TO_FILE)))
        raise ValueError(f"Unsupported electron impact species {species!r}. Supported keys: {supported}") from exc


@lru_cache(maxsize=None)
def _load_cross_section(filename: str) -> tuple[np.ndarray, np.ndarray]:
    path = files("py_ionization.electron_impact").joinpath("data", filename)
    rows = []
    with path.open("r", encoding="utf-8", errors="ignore") as handle:
        for line in handle:
            stripped = line.strip()
            if not stripped or stripped.startswith("#"):
                continue
            parts = stripped.split(",") if "," in stripped else stripped.split()
            try:
                rows.append([float(part) for part in parts])
            except ValueError:
                continue
    if not rows:
        raise ValueError(f"No numeric data lines found in {path}.")
    data = np.asarray(rows, dtype=float)
    column = 1 if data.shape[1] == 2 else 2
    return read_x1e16_table(path, column=column)


def electron_impact_cross_section(energy_eV, species: str = "H"):
    """Interpolate an electron impact ionization cross section.

    Parameters
    ----------
    energy_eV
        Electron energy in eV. Scalar and array inputs are accepted.
    species
        Neutral target species, for example ``"H"``, ``"O"``, ``"O2"``,
        ``"CO2"``, ``"N2"``, ``"Na"``, or ``"Mg"``.

    Returns
    -------
    sigma_cm2
        Ionization cross section in cm^2. Values outside the tabulated energy
        range are returned as NaN.
    """
    energy_arr = np.asarray(energy_eV, dtype=float)
    scalar_input = energy_arr.ndim == 0
    grid_eV, sigma_cm2 = _load_cross_section(_data_file(species))
    out = np.interp(energy_arr, grid_eV, sigma_cm2, left=np.nan, right=np.nan)
    if scalar_input:
        return float(np.asarray(out))
    return out


ei_cross_section = electron_impact_cross_section
cross_section = electron_impact_cross_section

__all__ = ["electron_impact_cross_section", "ei_cross_section", "cross_section"]

File: matlab/+py_ionization/+electron_impact/cross_section.m
function sigma_cm2 = cross_section(energy_eV, species)
%CROSS_SECTION Interpolate an electron impact ionization cross section.
%
% sigma_cm2 = py_ionization.electron_impact.cross_section(energy_eV)
% sigma_cm2 = py_ionization.electron_impact.cross_section(energy_eV, species)
%
% Inputs
%   energy_eV  Electron energy in eV.
%   species    Neutral target species, for example H, O, O2, CO2, N2, Na.
%
% Output
%   sigma_cm2  Cross section in cm^2. Values outside the tabulated energy
%              range are returned as NaN.

if nargin < 2
    species = "H";
end

filename = species_to_file(species);
path = py_ionization.data_path("electron_impact", filename);
data = read_numeric_table(path);
if size(data, 2) == 2
    column_index = 2;
else
    column_index = 3;
end
[grid_eV, sigma_table] = py_ionization.read_x1e16_table(path, column_index);

input_size = size(energy_eV);
energy_vec = double(energy_eV(:));
sigma_vec = interp1(grid_eV, sigma_table, energy_vec, "linear", nan);
sigma_cm2 = reshape(sigma_vec, input_size);
end

function filename = species_to_file(species)
key = lower(regexprep(string(species), "\s+", ""));

switch key
    case {"h", "h+"}
        filename = "electron_impact_H.txt";
    case {"c", "c+"}
        filename = "electron_impact_C.txt";
    case {"o", "o+"}
        filename = "electron_impact_O.txt";
    case {"o2", "o2+"}
        filename = "electron_impact_O2.txt";
    case {"co", "co+"}
        filename = "electron_impact_CO.txt";
    case {"n2", "n2+"}
        filename = "electron_impact_N2.txt";
    case {"co2", "co2+"}
        filename = "electron_impact_CO2.txt";
    case {"ar", "ar+"}
        filename = "electron_impact_Ar.txt";
    case {"na", "na+"}
        filename = "electron_impact_Na.txt";
    case {"mg", "mg+"}
        filename = "electron_impact_Mg.txt";
    case {"al", "al+"}
        filename = "electron_impact_Al.txt";
    case {"si", "si+"}
        filename = "electron_impact_Si.txt";
    case {"k", "k+"}
        filename = "electron_impact_K.txt";
    case {"ca", "ca+"}
        filename = "electron_impact_Ca.txt";
    case {"ti", "ti+"}
        filename = "electron_impact_Ti.txt";
    case {"fe", "fe+"}
        filename = "electron_impact_Fe.txt";
    otherwise
        error("py_ionization:electron_impact:UnsupportedSpecies", ...
            "Unsupported electron impact species: %s", species);
end
end

function data = read_numeric_table(path)
fid = fopen(path, "r");
if fid < 0
    error("py_ionization:electron_impact:OpenFailed", ...
        "Could not open file: %s", path);
end
cleanup = onCleanup(@() fclose(fid));
rows = {};
while true
    line = fgetl(fid);
    if ~ischar(line)
        break
    end
    line = strtrim(line);
    if line == "" || startsWith(line, "#")
        continue
    end
    values = sscanf(strrep(line, ",", " "), "%f").';
    if ~isempty(values)
        rows{end + 1, 1} = values; %#ok<AGROW>
    end
end
n_col = max(cellfun(@numel, rows));
data = nan(numel(rows), n_col);
for i = 1:numel(rows)
    values = rows{i};
    data(i, 1:numel(values)) = values;
end
end

